gobblin-data-management/src/main/java/gobblin/data/management/copy/entities/CommitStepDB.java

```diff
@@ -0,0 +1,206 @@
+   * @throws IOException for SQL errors.
+   * @throws IllegalStateException if a {@link CommitStep} with that key already exists.
+   */
+  public void put(String key, CommitStep step) throws IOException {
+    String serializedStep = CopyEntity.GSON.toJson(step);
+    try (PreparedStatement statement = this.conn.prepareStatement(INSERT_STEP_STATEMENT)) {
+      statement.setString(1, key);
+      statement.setAsciiStream(2, new ByteArrayInputStream(serializedStep.getBytes(Charsets.UTF_8)));
+      statement.execute();
+      this.conn.commit();
+    } catch (SQLIntegrityConstraintViolationException se) {
+      String existingStep = getRaw(key);
+      if (!serializedStep.equals(existingStep)) {
+        throw new IllegalStateException(
+            String.format("A step with key %s is already present in the commit step DB.", key));
+      }
+    } catch (SQLException se) {
+      throw new IOException(se);
+    }
+  }
+
+  /**
+   * Get the {@link CommitStep} stored under this key.
+   * @param key
+   * @return {@link CommitStep} stored under input key.
+   * @throws NoSuchElementException if no {@link CommitStep} with that key exists.
+   * @throws IOException for SQL errors.
+   */
+  public CommitStep get(String key) throws IOException {
+    return CopyEntity.GSON.fromJson(getRaw(key), CommitStep.class);
+  }
+
+  private String getRaw(String key) throws IOException {
+    try (PreparedStatement statement = this.conn.prepareStatement(GET_STEP_STATEMENT)) {
+      statement.setString(1, key);
+      ResultSet rs = statement.executeQuery();
+      this.conn.commit();
+
+      if (!rs.next()) {
+        throw new NoSuchElementException("No step with key " + key);
+      }
+
+      InputStream is = rs.getAsciiStream(1);
+      return IOUtils.toString(is, Charsets.UTF_8);
+    } catch (SQLException se) {
+      throw new IOException(se);
+    }
+  }
+
+  /**
+   * Closes the connection to the DB.
+   * @throws IOException
+   */
+  @Override
+  public void close()
+      throws IOException {
+    try {
+      this.conn.close();
+    } catch (SQLException sqle) {
+      log.error("Failed to close SQL connection.");
+      printSQLException(sqle);
+    }
+  }
+
+  /**
+   * Prints details of an SQLException chain to <code>System.err</code>.
+   * Details included are SQL State, Error code, Exception message.
+   *
+   * @param e the SQLException from which to print details.
+   */
+  public static void printSQLException(SQLException e) {
+    // Unwraps the entire exception chain to unveil the real cause of the
+    // Exception.
+    while (e != null) {
+      System.err.println("\n----- SQLException -----");
+      System.err.println("  SQL State:  " + e.getSQLState());
+      System.err.println("  Error Code: " + e.getErrorCode());
+      System.err.println("  Message:    " + e.getMessage());
+      // for stack traces, refer to derby.log or uncomment this:
+      //e.printStackTrace(System.err);
+      e = e.getNextException();
```

Comment:
Are you trying to print all the "caused by" exceptions?
Did you check if https://commons.apache.org/proper/commons-lang/javadocs/api-3.1/org/apache/commons/lang3/exception/ExceptionUtils.html has what you need?